### ISSUE TITLE
White blank after AD ENDED event - Player crashes with vast 4.0+ IOS

### ISSUE THREAD

**Issue by damjan25** [googleads/videojs-ima#931]
Device: Mobile IOS(Iphone) versions 13.6 and 14.0.1
Browser: Chrome (also on safari)
"@types/video.js": "^7.2.16" (tried with 7.3.10 and it didnt work)
"videojs-ima": "^1.7.2" (tried with 1.8.1 and it didnt work)

I am using videojs with ima sdk plugin.
How can I set listener for MEDIA_ERR_SRC_NOT_SUPPORTED error ? So I can then at least close the whole player, because right now it just shows the black screen and mentioned error and the close button doesnt work.
I tried with this, but it does not work.
player.on('MEDIA_ERR_SRC_NOT_SUPPORTED', function () {});

I also tried with player.on('error',...) but it was not triggered and I couldnt close the player (close button is not working after the error)

Is it possible to handle to catch this error and close player somehow ?

Steps to reproduce:
- ad 2 VAST-s first one with good/valid ad and second one with bad ad
- after first one is done playing, the second one sometimes crashes the whole player and you cant close it

I was able to reproduce this on PC chrome by triggering player.play() on event adEnded

I can also provide bad ad sample via slack

![image](https://user-images.githubusercontent.com/60606529/90193241-d54ad080-ddc4-11ea-865e-426fc4923818.png)

**Comment by Kiro705** [comment]
Hello @damjan25 ,

It looks like the error you are getting is coming from the videoJS and not the IMA SDK. I was able to find this issue related to videoJS error handling (https://github.com/videojs/video.js/issues/6280).

Let me know if that helps, and feel free to reply here if you have questions.

Thank you,
Jackson
IMA SDK DevRel

**Comment by damjan25** [comment]
@Kiro705  looks like this happens, when there are 2 instances of videojs with ima sdk plugin on same website.
But it is really weird that it happens only on IOS

**Comment by Kiro705** [comment]
Hello @damjan25 ,

Having multiple instances of the plugin running is not currently a supported use case. I would recommend looking for a way to avoid this case, as it is not something IMA or this plugin currently test against.

Thank you,
Jackson
IMA SDK DevRel

**Comment by damjan25** [comment]
The weird thing is that it works perfectly on desktop :P

**Comment by damjan25** [comment]
@Kiro705  looks like vast 4.0 is not supported on ios 
If you want I can send you vast tag on videojs-slack :)

**Comment by Kiro705** [comment]
If instead, could you send the ad tag to my email address, jsui@google.com. I can test the tag from there and raise the appropriate issues.

**Comment by Kiro705** [comment]
Hello @damjan25 ,

Thank you for privately sharing the ad tag. When I tested the tag in the [Video Suite Inspector (VSI)](https://developers.google.com/interactive-media-ads/docs/sdks/html5/client-side/vastinspector), it returned `Ad error: AdError 303: No Ads VAST response after one or more Wrappers`. I wanted to confirm with you if this is the same behavior you are seeing. 

To note, the VSI is an implementation of the IMA SDK, but does not use the videoJS-IMA plugin. However, we do use it as our baseline for support. 

I was able to play the ad-tag in the [advanced example of videoJS-IMA plugin ](https://googleads.github.io/videojs-ima/examples/advanced/) which is a little confusing why that would differ from the behavior seen the in VSI. The ad played on desktop, as well as an emulated iPhoneX (from the Chrome devtools). Is this the same behavior you are seeing?

Thank you,
Jackson
IMA SDK DevRel

**Comment by damjan25** [comment]
desktop chrome: works 
android (real device): works
ios (real device): does not work, in VSI I get adError 303 but in my videojs-ima integration it breaks whole player (I tried to listen to every possible error event but it is just not catching the error)

**Comment by damjan25** [comment]
@Kiro705 Anything new on this ? We are still getting a lot of reports that on IoS users are getting this blank error.
It works on emulated Iphone but it does not work on real device. Is there different error handling on ios and android ?

**Comment by damjan25** [comment]
@Kiro705 If I mock userAgent on IOS Iphone real device, to the one bellow, it works perfectly :yum: 

      let userAgent = 'Mozilla/5.0 (Linux; Android 10; Mi MIX 2S) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.125 Mobile Safari/537.36';

      window.navigator.__defineGetter__('userAgent', function(){
        return userAgent;
      });

**Comment by Kiro705** [comment]
Hello @damjan25 ,

I think the issue is related to a malformed ad-tag that is being requested specifically for iOS devices. When I was able to capture the VAST for that ad, I was able to reproduce the 303 error on all platforms. 

I would recommend taking a look at your ads manager settings to confirm all ad-tags for the url you shared are properly formatted.

**Comment by damjan25** [comment]
@Kiro705 problem is that it is happening with a lot of ad-tags and on many websites. I sent you mail(subject: damjan25 - github - ios vast4 not working) with live example if it helps.

**Comment by Kiro705** [comment]
Hello @damjan25 ,

Thank you for sharing the live site. On the site I was able to see ad playback occur once, and then no content playing after. On desktop, I see and ad playing, then no content, but another ad playing after that (so maybe another bug?). To note, the IMA's intended use is for ads to be played with some sort of interactive media, and not as stand-alone ads.

I have tested the ad on a native iOS device using the [videojs-ima advanced sample](https://googleads.github.io/videojs-ima/examples/advanced/) and saw ad playback followed by content playback as expected. See attached images.
![ad_playback](https://user-images.githubusercontent.com/14110467/94287422-46cd8100-ff24-11ea-845a-048c0f90f533.png)

![content_playback](https://user-images.githubusercontent.com/14110467/94287436-4a610800-ff24-11ea-888d-e3bdd05b9b80.png)

Would it be possible to share a code sample that is able to reproduce the issue?

**Comment by damjan25** [comment]
@Kiro705 to reprocude this with advanced example ( https://github.com/googleads/videojs-ima/tree/master/examples/advanced)
you can do this:
- First to check if it handles ADSERROR:
1. Advanced exampe/ads.js => init func =>  line 71 just before calling this.player.ima.initializeAdDisplayContainer();
paste in listener for adserror:
```js
this.player.on("adserror", () => {
      this.log(window.navigator.userAgent);
      this.log('adserror listener triggered');
    });
```

on desktop you will see in textarea output => "adserror listener triggered'"
on IOS you wont see anything

- if you want to make it handle the adserror then:

1. in Advanced exampe/index.html on line 19, right under <head> starting tag paste in this code:
```html
<script type='text/javascript'>
      let userAgent = 'Mozilla/5.0 (Linux; Android 10; Mi MIX 2S) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.125 Mobile Safari/537.36';

      window.navigator.__defineGetter__('userAgent', function(){
        return userAgent;
      });
    </script>
```
And the adserror will be handled properly even on IOS :)

**Comment by damjan25** [comment]
@Kiro705 looks like IMA does not support vast 4.0+ on IOS and this crashes the player.

```xml
<VAST version="4.1" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns="http://www.iab.com/VAST">
  <Ad id="20001" sequence="1" conditionalAd="false">
    <InLine>
      <AdSystem version="4.1">iabtechlab</AdSystem>
      <Error><![CDATA[https://example.com/error]]></Error>
      <Extensions>
        <Extension type="iab-Count">
          <total_available>
            <![CDATA[ 2 ]]>
          </total_available>
        </Extension>
      </Extensions>
      <Impression id="Impression-ID"><![CDATA[https://example.com/track/impression]]></Impression>
      <Pricing model="cpm" currency="USD">
        <![CDATA[ 25.00 ]]>
      </Pricing>
      <AdServingId>a532d16d-4d7f-4440-bd29-2ec0e693fc80</AdServingId>
      <AdTitle>iabtechlab video ad</AdTitle>
      <AdVerifications>
        <Verification>
          <JavaScriptResource>
            <![CDATA[https://verificationcompany1.com/verification_script1.js]]>
          </JavaScriptResource>
        </Verification>
        <Verification>
          <JavaScriptResource>
            <![CDATA[https://verificationcompany.com/untrusted.js]]>
          </JavaScriptResource>
        </Verification>
      </AdVerifications>
      <Advertiser>IAB Sample Company</Advertiser>
      <Category authority="foo" />
      <Creatives>
        <Creative id="5480" sequence="1" adId="2447226">
          <UniversalAdId idRegistry="Ad-ID" >8465</UniversalAdId>
          <Linear>
            <TrackingEvents>
              <Tracking event="start" ><![CDATA[https://example.com/tracking/start]]></Tracking>
              <Tracking event="progress" offset="00:00:10"><![CDATA[http://example.com/tracking/progress-10]]></Tracking>
              <Tracking event="firstQuartile"><![CDATA[https://example.com/tracking/firstQuartile]]></Tracking>
              <Tracking event="midpoint"><![CDATA[https://example.com/tracking/midpoint]]></Tracking>
              <Tracking event="thirdQuartile"><![CDATA[https://example.com/tracking/thirdQuartile]]></Tracking>
              <Tracking event="complete"><![CDATA[https://example.com/tracking/complete]]></Tracking>
            </TrackingEvents>
            <Duration>00:00:16</Duration>
            <MediaFiles>
              <MediaFile id="5241" delivery="progressive" type="video/mp4" bitrate="2000" width="1280" height="720" minBitrate="1500" maxBitrate="2500" scalable="1" maintainAspectRatio="1" codec="H.264">
                <![CDATA[https://iab-publicfiles.s3.amazonaws.com/vast/VAST-4.0-Short-Intro.mp4]]>
              </MediaFile>
              <MediaFile id="5244" delivery="progressive" type="video/mp4" bitrate="1000" width="854" height="480" minBitrate="700" maxBitrate="1500" scalable="1" maintainAspectRatio="1" codec="H.264">
                <![CDATA[https://iab-publicfiles.s3.amazonaws.com/vast/VAST-4.0-Short-Intro-mid-resolution.mp4]]>
              </MediaFile>
              <MediaFile id="5246" delivery="progressive" type="video/mp4" bitrate="600" width="640" height="360" minBitrate="500" maxBitrate="700" scalable="1" maintainAspectRatio="1" codec="H.264">
                <![CDATA[https://iab-publicfiles.s3.amazonaws.com/vast/VAST-4.0-Short-Intro-low-resolution.mp4]]>
              </MediaFile>
            </MediaFiles>
            <VideoClicks>
              <ClickThrough id="blog">
                <![CDATA[https://iabtechlab.com]]>
              </ClickThrough>
            </VideoClicks>
          </Linear>
        </Creative>
      </Creatives>
    </InLine>
  </Ad>
</VAST>
```

**Comment by Kiro705** [comment]
Hello @damjan25 ,

I am still having trouble reproducing the player crash you are describing.

When I use the VAST provided in the [hosted advanced example](https://googleads.github.io/videojs-ima/examples/advanced/), I see successful ad and content play back on both desktop and iOS mobile web.

In addition, when I test with the changes you described, I see successful playback on desktop.

Can you clarify, if the vast you provided is expected to work on iOS mobile web for the [hosted advanced example](https://googleads.github.io/videojs-ima/examples/advanced/)?

Also, can you clarify which ENDED event you are looking to. My understanding is the IMA does not have an ENDED event ([list of IMA events](https://developers.google.com/interactive-media-ads/docs/sdks/html5/client-side/reference/js/google.ima.AdEvent#.Type))

**Comment by damjan25** [comment]
@Kiro705 But do you see on IOS(real device) adserror getting triggered ? 

1. Vast provided is supposed to play ad on plays the video on desktop and android. For some reason it does not work on IOS and ADSERROR is not being handled.
2. I was talking about event ENDED from videojs

**Comment by Kiro705** [comment]
Hi @damjan25 ,

I was able to reproduce the adserror event on iOS, but did not see a player crash. The events I saw were as follows.

1. After init function, this error occured: `Ad error: AdError 400: There was an error playing the video ad. Caused by: AdError 1205: The browser prevented playback initiated without user interaction.` This triggered the `adserror` event.
2. Ad did not play
3. Video content played
4. `ended` event occurred.
5. Player did not crash and I was able to replay the video.

I can look into the error and why it is occurring curing, since I did interact with the player. However, I still was unable to reproduce the player crash.
